wifi: Add full AP mode support (using hostapd)

nxf58150 opened the pull request:
Added support of setting up AP with hostapd. 
Once flag CONFIG_WIFI_NM_HOSTAPD_AP is enabled, softAP related commands in l2 wifi_shell.c will be handled by hostapd.
Increase max value of CONFIG_WIFI_SHELL_MAX_AP_STA from 5 to 8. The softAP of NXP wifi chip can support up to 8 stations.

nxf58150 committed:
modules: hostap: Add support of setting up softAP with hostapd

Added new flag CONFIG_WIFI_NM_HOSTAPD_AP for hostapd support. Once this
flag is enabled, softAP will be setup by hostapd. Both wpa_supplicant
and hostapd uses same task and eloop.
Included necessary hostapd files when compiling wifi samples if
CONFIG_WIFI_NM_HOSTAPD_AP is enabled. Added hostapd support for all
softAP command of L2 wifi shell commands.

Signed-off-by: Hui Bai <hui.bai@nxp.com>

nxf58150 committed:
net: l2: wifi: Increase max count of connected STA

Increase max count of CONFIG_WIFI_SHELL_MAX_AP_STA from 5 to 8. The
SoftAP of NXP wifi chip can support up to 8 stations.

Signed-off-by: Hui Bai <hui.bai@nxp.com>

nxf58150 committed:
modules: hostap: Supports link mode of legacy STA.

Link mode shows unknown when legacy (A or bg only) device connects to
APUT. Set the link mode to WIFI_2 when the host freq over 4000 and set
link mode to WIFI_3 when the host freq over 2000.

Signed-off-by: Hui Bai <hui.bai@nxp.com>

nxf58150 committed:
net: l2: wifi: Fix issue command wifi connect fail.

Increase wifi connect input parameters max count to 13. Previous count
7 is not enough if other security type is supported.
When enabling softAP, the parameter cnx_params in cmd_wifi_ap_enable()
is with static key word. Then the parameter will always save
configurations of last time. Remove static keyword to eliminate effects
of configs from last tim and do memset before setting up softAP.

Signed-off-by: Hui Bai <hui.bai@nxp.com>

jukkar commented on modules/hostap/src/supp_main.c:
```diff
@@ -523,6 +585,456 @@ static int register_supplicant_event_socket(struct supplicant_context *ctx)
+	os_strlcpy(conf->bss[0]->iface, ifname, sizeof(conf->bss[0]->iface));
+	bss->logger_stdout_level = HOSTAPD_LEVEL_INFO;
+	bss->logger_stdout       = 0xffff;
+	bss->nas_identifier      = os_strdup("ap.example.com");
+	bss->eap_sim_db          = os_strdup("unix:/tmp/hlr_auc_gw.sock");
+	os_memcpy(conf->country, "US ", 3);
```
The domain socket is something that is not needed in zephyr.
Also the hardcoded values look wrong.

nxf58150 replied:
Removed line 845.

jukkar commented on modules/hostap/src/supp_main.c:
```diff
@@ -523,6 +585,456 @@ static int register_supplicant_event_socket(struct supplicant_context *ctx)
+	iface = net_if_get_wifi_uap();
+	/* Allocate and parse configuration for full interface files */
+	dev = net_if_get_device(iface);
+	strncpy(ifname, dev->name, IFNAMSIZ);
```
We should not use device name as an interface name.

nxf58150 replied:
Updated.

jukkar commented:
Isn't this depending on https://github.com/zephyrproject-rtos/hostap/pull/15, the west.yml should be updated too in that case.

krish2718 commented on modules/hostap/Kconfig:
```diff
@@ -171,6 +175,9 @@ config WIFI_NM_WPA_SUPPLICANT_WPA3
 	bool "AP mode support"
 
+config WIFI_NM_HOSTAPD_AP
+	bool "AP mode hostapd support"
```
`hostapd based full AP mode` and we can fix `WIFI_NM_WPA_SUPPLICANT_AP` to ` SoftAP mode support based on WPA supplicant`

nxf58150 replied:
Updated.

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
+{
+	int ret = 0;
+
+	if (!hostapd_cli_cmd_v("set ssid %s", params->ssid)) {
+		goto out;
+	}
```
you can use a macro do the NULL check + goto out, like we did for WPA supplicant.

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
+	return ret;
+}
+
+int hapd_config_network(struct hostapd_iface *iface,
```
IIUC, the commands itself are common b/w SoftAP and FullAP, so, may be if you write a wrapper to choose wpa_cli_cmd_v vs hostapd_cli_cmd_v, we can reuse the code?

nxf58150 replied:
Do you mean reuse the code of configuring network? I think the commands between softAP and fullAP are different.

krish2718 replied:
Yes.

> I think the commands between softAP and fullAP are different.

Ok, let me check the differences in commands then, thanks.

krish2718 reviewed (commented):
I don't see the dynamic interfaces support for HOSTAPD_AP i.e., monitoring interface state events and accordingly stop/start using the interface, is this planned?

krish2718 commented:
Should we also extend `wifi_nm_iface_type` with `FAP` i.e., `FULL AP`?

* `SoftAP` -> WPA supplicant
* `FullAP` -> hostapd

nxf58150 commented on modules/hostap/src/supp_events.c:
```diff
@@ -291,21 +311,27 @@ int supplicant_send_wifi_mgmt_ap_sta_event(void *ctx,
 	struct sta_info *sta = data;
-	struct wpa_supplicant *wpa_s = ctx;
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+	struct hostapd_iface *iface = ctx;
+	char *ifname = iface->bss[0]->conf->iface;
```
How about use name `"ap_ctx"` instead of `"iface"` here?

nxf58150 replied:
Renamed to `ap_ctx` here.

nxf58150 commented:
> I don't see the dynamic interfaces support for HOSTAPD_AP i.e., monitoring interface state events and accordingly stop/start using the interface, is this planned?

Currently, we are not using dynamic interface support on hostapd. We need time to do research on this. How about we keep it like this for now? Will do code enhancement in the future.

krish2718 commented:
> > I don't see the dynamic interfaces support for HOSTAPD_AP i.e., monitoring interface state events and accordingly stop/start using the interface, is this planned?
> 
> Currently, we are not using dynamic interface support on hostapd. We need time to do research on this. How about we keep it like this for now? Will do code enhancement in the future.

But other drivers might still use it (Nordic driver uses it, but still not upstream). May be add a Kconfig dependency on `depends on !WIFI_NM_WPA_SUPPLICANT_INF_MON`?

krish2718 commented:
> > Should we also extend `wifi_nm_iface_type` with `FAP` i.e., `FULL AP`?
> > 
> > * `SoftAP` -> WPA supplicant
> > * `FullAP` -> hostapd
> 
> Setting up AP with hostapd or wpa_supplicant is transparent for user so I think extending `wifi_nm_iface_type` maybe not a necessary task. By checking the type, user will know the interface is AP, no matter it is SoftAP or FullAP. If we don't extend it, will it affect setting up AP with wpa_supplicant?

Agreed, in that case let's rename SAP ->AP and hide Soft/Full.

nxf58150 commented:
> Isn't this depend on [zephyrproject-rtos/hostap#15](https://github.com/zephyrproject-rtos/hostap/pull/15), the west.yml should be updated too in that case.

Updated west.yml.

nxf58150 commented:
> Should we also extend `wifi_nm_iface_type` with `FAP` i.e., `FULL AP`?
> 
> * `SoftAP` -> WPA supplicant
> * `FullAP` -> hostapd

As discussed, will upload another PR to rename iface type from SAP to AP.

nxf58150 commented:
> HOSTAPD_AP

The station interface of our chip is still using WIFI_NM_WPA_SUPPLICANT_INF_MON so adding dependency of !WIFI_NM_WPA_SUPPLICANT_INF_MON for HOSTAPD_AP will have side effect.  The hostapd is not using dynamic interface support currently but it doesn't affect the functionality of WIFI_NM_WPA_SUPPLICANT_INF_MON.

jerome-pouiller commented on modules/hostap/src/supp_api.c:
```diff
@@ -951,11 +1187,36 @@ int supplicant_ap_enable(const struct device *dev,
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+	iface = get_hostapd_handle(dev);
+	if (!iface) {
+		ret = -1;
```
`ENOENT`?

nxf58150 replied:
Modified `-1` to `ENOENT`.

jerome-pouiller commented on modules/hostap/src/supp_api.h:
```diff
@@ -138,6 +138,17 @@ int supplicant_filter(const struct device *dev, struct wifi_filter_info *filter)
+ * @return 0 for OK; -1 for ERROR
+ */
+int hapd_state(const struct device *dev, int *state);
+#endif
```
I suggest to provide the same public API whatever the value of `CONFIG_WIFI_NM_HOSTAPD_AP`:
```
#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
int hapd_state(const struct device *dev, int *state);
#else
static inline int hapd_state(const struct device *dev, int *state) {
    return -EINVAL;
}
#endif
```

Thus the caller does not have to add `#ifdef` around the call to this function.

nxf58150 replied:
Updated.

jerome-pouiller commented on modules/hostap/src/supp_main.c:
```diff
@@ -523,6 +585,455 @@ static int register_supplicant_event_socket(struct supplicant_context *ctx)
+	conf->acs            = conf->channel == 0;
+#ifdef CONFIG_ACS
+	conf->acs_num_scans  = 1;
+#endif /* CONFIG_ACS */
```
Don't Repeat Yourself. If the condition is obvious, I suggest to not repeat it after the `#endif`.

nxf58150 replied:
Removed comment after `#endif`.

jerome-pouiller commented on modules/hostap/src/supp_main.c:
```diff
@@ -523,6 +585,455 @@ static int register_supplicant_event_socket(struct supplicant_context *ctx)
+	interfaces->config_read_cb     = hostapd_config_read2;
+	interfaces->for_each_interface = hostapd_for_each_interface;
+	interfaces->ctrl_iface_init    = NULL;
+	interfaces->ctrl_iface_deinit  = NULL;
```
This assignment is useless since you already called `memset()` (Probably you think it make sense to assign all the members of `interfaces`. However content of `interfaces` will probably change and the way the members are initialized will look illogical).

nxf58150 replied:
Removed unnecessary assignment.

jerome-pouiller commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1295,9 +1295,10 @@ static int cmd_wifi_ap_enable(const struct shell *sh, size_t argc,
 {
 	struct net_if *iface = net_if_get_wifi_sap();
-	static struct wifi_connect_req_params cnx_params;
+	struct wifi_connect_req_params cnx_params;
```
This patch change the lifetime of `cnx_params`. Can you confirm you have checked address of this variable is never copied somewhere else?

nxf58150 replied:
With `"static"`, this parameter will carry configurations from last time. For example, I first enable AP of WPA-PSK mode and set psk in cnx_params. Later, I disabled this AP and start it as open mode. There will be warning log `"Passphrase provided without security configuration"`. 
I think it is meaningless to use "static" because AP configurations might be different for each time. 
This parameter only used in this API so I simply removed the keywork "static" and do memset of this parameter.

jerome-pouiller reviewed (commented):
Sorry the late review.

Thank you for your effort. I hope my comments are not too late.

nxf58150 commented on modules/hostap/src/supp_api.c:
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
+	iface = get_hostapd_handle(dev);
+	if (!iface) {
+		wpa_printf(MSG_ERROR, "Device %s not found", dev->name);
+		ret = -1;
```
Modified `-1` to `ENOENT`.

nxf58150 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1295,9 +1295,10 @@ static int cmd_wifi_ap_enable(const struct shell *sh, size_t argc,
+	struct wifi_connect_req_params cnx_params;
 	int ret;
 
+	memset(&cnx_params, 0, sizeof(struct wifi_connect_req_params));
```
Updated to struct `wifi_connect_req_params cnx_params = {0};`

nxf58150 commented on modules/hostap/src/supp_api.c:
```diff
@@ -951,11 +1187,36 @@ int supplicant_ap_enable(const struct device *dev,
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+	struct hostapd_iface *iface;
+	int ret = 0;
+#else
```
Vendors can choose to setup AP with hostapd by enabling this flag. Or, the AP will be setup by wpa_supplicant. 
Also, I found that there are many files in Zephyr has `#ifdef` in .c files, such as net_if.c, tcp.c. So I think it might be ok if I put `#ifdef` here?

jerome-pouiller replied:
Obviously, conditional compilation is needed. However, I am afraid this file become a nightmare to maintain with `#ifdef` inside.

I suggest to split this file in two (or three). Then, rely on CMakeLists.txt (using `zephyr_library_sources_ifdef()`) to manage conditional compilation.

krish2718 replied:
Sure, got it, may be `hapd_api.c` and then move all hostapd related code to that file.

gannicus-gj replied:
Hi @krish2718 
if split supp_api.c to hapd_api.c, do you think supp_main.c and supp_events.c needs same split?

the reason I said a lot work to do is on our internal code base, because  following feature which already finished, such as: uap enterpise,  ap dpp are all done based on supp_api.c (none split file),
so there will be a lot of rebase work to do in our internal code base

do you think it is possible that we first finish upstream all our PRs in this none split file, and finally move to a split file
for supp_api.c, supp_events.c and supp_main.c?

krish2718 replied:
Fine with me, we can add an improvement in GH to track.

gannicus-gj replied:
Hi @jerome-pouiller , do you agree with our above plan plan? could you pls take time to comment?
thanks in advance

gannicus-gj replied:
Hi @jerome-pouiller 
any feedback/comments on following plan ?
 if split supp_api.c to hapd_api.c, do you think supp_main.c and supp_events.c needs same split?

the reason I said a lot work to do is on our internal code base, because target to upstream features which already finished, such as: uap enterpise, ap dpp are all done based on supp_api.c (none split file), so there will be a lot of rebase work to do in our internal code base

do you think it is possible that we first finish upstream all our PRs in this none split file, and finally move to a split file for supp_api.c, supp_events.c and supp_main.c in one Pull Request ?

nxf58150 commented on modules/hostap/src/supp_api.c:
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
+		if (!hostapd_cli_cmd_v("set ieee80211w %d", params->mfp)) {
+			goto out;
+		}
+	}
```
Removed the if condition check for this parameter. Now, the `"set ieee80211w"` will always run here.

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
+			if (!hostapd_cli_cmd_v("set wpa_key_mgmt WPA-PSK")) {
+				goto out;
+			}
+			if (!hostapd_cli_cmd_v("set wpa_passphrase %s", params->psk)) {
```
Yes, in `supp_api.c` we always use `\"%s\"` to handle the spaces.

nxf58150 committed:
manifest: Update hostap to latest

Update hostap repo to latest in west.yml

Signed-off-by: Hui Bai <hui.bai@nxp.com>

jukkar commented:
There is quite weird error in CI
```
CONFLICT (content): Merge conflict in west.yml
error: could not apply 971c7cdaba8... manifest: Update hostap to latest
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 971c7cdaba8... manifest: Update hostap to latest
```
Could you rebase this PR against latest upstream, and force push, perhaps it solves this issue.

jukkar commented on modules/hostap/src/supp_api.h:
```diff
@@ -190,6 +190,21 @@ int supplicant_get_wifi_conn_params(const struct device *dev,
+ */
+int hapd_state(const struct device *dev, int *state);
+#else
+int hapd_state(const struct device *dev, int *state) {
```
This function should be static inline as suggested by @jerome-pouiller

nxf58150 replied:
Updated. Thanks!

jukkar reviewed (dismissed):
LGTM

jukkar commented:
@jerome-pouiller & @krish2718 please take a look

krish2718 commented on modules/hostap/CMakeLists.txt:
```diff
@@ -209,11 +210,31 @@ zephyr_library_compile_definitions_ifdef(CONFIG_WIFI_NM_WPA_SUPPLICANT_AP
 	CONFIG_NO_VLAN
 	CONFIG_NO_ACCOUNTING
 	NEED_AP_MLME
+	CONFIG_IEEE80211AC
```
why is this needed? As this is independent to how the AP is being setup, it should be in a separate commit.

nxf58150 replied:
Our chip supports 11AC(both STA and AP) and is enabled by default. Without this flag, 11AC related source code won't be compiled and set conf->ieee80211ac is useless.

krish2718 replied:
ok, can you please put this in a separate commit.

nxf58150 replied:
Updated.

nxf58150 committed:
modules: hostap: Enable CONFIG_IEEE80211AC flag for hostapd support.

Enable CONFIG_IEEE80211AC flag for 11AC support when setting up
AP with hostapd.

Signed-off-by: Hui Bai <hui.bai@nxp.com>

jukkar commented:
@nxf58150 the hostap PR is now merged, please update manifest file and re-submit this one

nxf58150 commented:
> @nxf58150 the hostap PR is now merged, please update manifest file and re-submit this one

Updated manifest file.
